cradle-core/src/main/java/com/exactpro/cradle/BookInfo.java

```diff
@@ -84,102 +148,247 @@ public Instant getCreated()
+	}
+
+	@SuppressWarnings("ResultOfMethodCallIgnored")
+	void invalidate() {
+		operateCache.invalidateAll();
+		randomAccessCache.invalidateAll();
+		long currentEpochDate = currentEpochDay();
+		operateCache.get(currentEpochDate - 1);
+		operateCache.get(currentEpochDate);
 	}
 	
-	
-	void removePage(PageId pageId)
-	{
-		PageInfo pageInfo = pages.remove(pageId);
-		if (pageInfo != null)
-			orderedPages.remove(pageInfo.getStarted());
+	void removePage(PageId pageId) {
+		long epochDate = getEpochDay(pageId.getStart());
+		long currentEpochDate = currentEpochDay();
+		LoadingCache<Long, IPageInterval> cache = currentEpochDate - epochDate < 2
+				? operateCache
+				: randomAccessCache;
+
+		IPageInterval pageInterval = cache.getIfPresent(epochDate);
+		if (pageInterval != null) {
+			pageInterval.remove(pageId);
+		}
 	}
 	
-	void addPage(PageInfo page)
-	{
-		pages.put(page.getId(), page);
-		orderedPages.put(page.getStarted(), page);
+	void addPage(PageInfo page) {
+		long epochDate = getEpochDay(page.getId().getStart());
+		long currentEpochDate = currentEpochDay();
+		LoadingCache<Long, IPageInterval> cache = currentEpochDate - epochDate < 2
+				? operateCache
+				: randomAccessCache;
+
+		cache.get(epochDate, key -> new PageInterval()).add(page);
+	}
+
+	@FunctionalInterface
+	public interface PagesLoader {
+		/**
+		 * @return page info collection from start to end date time. Both borders are included
+		 */
+		Collection<PageInfo> load(BookId bookId, Instant start, Instant end);
+	}
+
+	private IPageInterval getPageInterval(long epochDate) {
+		long currentEpochDate = currentEpochDay();
+		IPageInterval pageInterval = currentEpochDate - epochDate < 2
+				? operateCache.get(epochDate)
+				: randomAccessCache.get(epochDate);
+		return pageInterval != null
+				? pageInterval
+				: EMPTY_PAGE_INTERVAL;
+	}
+
+	private IPageInterval getPageInterval(Instant timestamp) {
+		long epochDate = getEpochDay(timestamp);
+		return getPageInterval(epochDate);
+	}
+
+	private @Nullable IPageInterval createPageInterval(Long epochDay) {
+		Instant start = toInstant(epochDay);
+		Instant end = start.plus(1, ChronoUnit.DAYS).minus(1, ChronoUnit.NANOS);
+		Collection<PageInfo> loaded = pagesLoader.load(id, start, end);
+		return loaded.isEmpty() ? null : create(loaded);
+	}
+
+	private static long currentEpochDay() {
+		return System.currentTimeMillis() / MILLISECONDS_IN_DAY;
+	}
+
+	private static long getEpochDay(Instant instant) {
+		return instant.getEpochSecond() / SECONDS_IN_DAY;
 	}
 
-	@Override
-	public boolean equals(Object o) {
-		if (this == o) return true;
-		if (o == null || getClass() != o.getClass()) return false;
-		BookInfo bookInfo = (BookInfo) o;
+	private static Instant toInstant(long epochDay) {
+		return  Instant.ofEpochSecond(epochDay * SECONDS_IN_DAY);
+	}
+
+	@FunctionalInterface
+	public interface PageLoader {
+		@Nullable
+		PageInfo load(BookId bookId);
+	}
+
+	private interface IPageInterval {
```

Comment:
I would think it should be something more similar to the Iterable interface. All the operations, correct me here if I am wrong, would look like:
1) Find the page for a certain start timestamp
2) Go from this page to the next/previous page until the page timestamp is less/greater than a certain end timestamp.

For me, it looks like this:
```java
interface IPageInterval {
  PageInfo get(PageId pageId)
  Iterator<PageInfo> getPages(Instant start, Instant end)
  void add(PageInfo page)
  void remove(PageId pageId)
}
```

Reply:
1. Do you mean use one sorted map and look for certain or nearest page by timestamp (current find function)?
2. I'm OK about that

Reply:
Yes, I would use the sorted map for that. Also, the sorted map provides us with convenient methods to get all data with keys greater or less than the provided one. This can be used to get to correct slice of the interval and  iterate over those pages